### PR TITLE
vehicle construction now informs about adjusted strength values

### LINKED ISSUE #32361: Vehicle part additional requirements are inaccurate if viewed with Bad Back
# Describe the bug

When removing/installing parts the requirements for strength (e.g. Lifting 1 or 10 Strength) are different if you have Bad Back, but still display old values but in red. I.e. I have 11 Strength and Bad Back and I want to remove a Heavy Duty Frame (req. Lifting 1 or Strength 10), it tells me I don't meet the requirement for 10 Str (shown in red) while I very clearly have 11 Strength.

# Steps To Reproduce

1. Make a Character with a Hacksaw, 10 Mechanics, 10 Strength and Bad Back
2. Spawn a Vehicle that uses Heavy Duty Frames.
3. Try to remove a Heavy Duty Frame and observe that it doesn't let you, while telling you that you need 10 Str (which you have).
4. Use Debug to remove Bad Back.
5. Observe that it is now possible to remove the Heavy Duty Frame.

# Expected behavior

For the Requirements listed to be the ones (and only ones) to remove parts.
Either make Bad Back not factor in here, change the requirement to be displayed differently depending on whether the player has Bad Back, or change the requirement displayed to be based on max weight capacity (and thus factor in bad back).

# Screenshots

Bug1
Bug2

# Versions and configuration

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on Windows 7
- Game Version: 0.D-5590-gabbd9f7 [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Icecoon's Arsenal [ew_pack],
    Makeshift Items Mod [makeshift],
    Medieval and Historic Classes and Shields [Medieval_Stuff],
    More Survival Tools [More_Survival_Tools],
    Mythological Replicas [nw_pack],
    Alternative Map Key [alt_map_key],
    Folding Parts pack [deoxymod],
    Prevent Zombie Revivication [no_reviving_zombies],
    Disable NPC Needs [no_npc_food],
    Aftershock [aftershock],
    Magiclysm [magiclysm],
    Vehicle Additions Pack [blazemod],
    Stats Through Kills [stats_through_kills],
    SpeedyDex [speedydex]
]

### PR DESCRIPTION
**Pull request by Jamuro-g**
#### Summary
Bugfixes "Vehicle construction ui informs about adjusted strength  due to strong, bad back"

#### Purpose of change

Installing/Removing vehicle part tasks already checked for those traits, but the ui never showed the adjusted values.
Players would have the required strength to lift a part, according to the interface but see it colored in red.
Now it checks if those traits were what caused the requirements to pass/fail and informs the player by how much their str value got adjusted for this check.

Fixes #32361

#### Describe the solution

Added a check for bad/good back traits to veh_interact::can_remove_part() & update_parts_requirements()
Those added messages only show up when the players not adjusted strength would have been either enough (bad back) or just under (good back) the requirements.

#### Testing

Debug a car in
Debug mutate in bad back/good back
Look at part installation/removal requirements

#### Additional context

How it loocks:
Strong Back:
image

Bad Back:
image
(old pic, it now says "hindered" instead of hurt, to avoid potential confusion)

### CODE REVIEW

File: src/veh_interact.cpp [hunk 1]
```diff
@@ -847,15 +847,34 @@ bool veh_interact::update_part_requirements()
         ok = false;
     }
 
+    std::string str_suffix;
+    int adjusted_strength = player_character.get_str();
+    if( player_character.has_trait( trait_id( "STRONGBACK" ) ) ) {
+        adjusted_strength *= 1.35;
+    } else if( player_character.has_trait( trait_id( "BADBACK" ) ) ) {
+        adjusted_strength /= 1.35;
+    }
```
Comment (BrettDong) [inline src/veh_interact.cpp:856]: I recommend extract the calculation of `STRONGBACK` or `BADBACK` adjusted strength to a helper member function of `player` class so there won't be duplicate code, and avoids inconsistency if someone changes this adjustment factor in one place in the future but forgets other two places.

File: src/veh_interact.cpp [hunk 2]
```diff
@@ -847,15 +847,34 @@ bool veh_interact::update_part_requirements()
...
+    adjusted_strength += player_character.get_lift_assist();
+
+    if( player_character.has_trait( trait_id( "STRONGBACK" ) ) && adjusted_strength >= str &&
+        ( player_character.get_str() + player_character.get_lift_assist() ) < str ) {
+        str_suffix = "(Strong Back helped with +" + std::to_string( static_cast<int>
+                     ( player_character.get_str() * 1.35 ) - player_character.get_str() ) + " strength)\n";
```
Comment (BrettDong) [inline src/veh_interact.cpp:862]:
```suggestion
        str_suffix = string_format( _( "(Strong Back helped with +%d strength)" ), static_cast<int>
                     ( player_character.get_str() * 1.35 ) - player_character.get_str() );
```

File: src/veh_interact.cpp (hunk 1 above)
Comment (Jamuro-g) [reply src/veh_interact.cpp]: good idea, should be updated now

File: src/veh_interact.cpp [hunk 3]
```diff
@@ -847,15 +847,29 @@ bool veh_interact::update_part_requirements()
...
     }
 
+    std::string str_suffix;
+    int lift_strength = player_character.get_lift_str();
+    int total_lift_strength = lift_strength + player_character.get_lift_assist();
+    int total_base_strength = player_character.get_str() + player_character.get_lift_assist();
+
+    if( player_character.has_trait( trait_id( "STRONGBACK" ) ) && total_lift_strength >= str &&
+        total_base_strength < str ) {
+        str_suffix = string_format( _( "(Strong Back helped with +%d strength)\n" ),
```
Comment (actual-nh) [inline src/veh_interact.cpp:857]:
```suggestion
        str_suffix = string_format( _( "(Strong Back helped, giving +%d strength)\n" ),
```

File: src/veh_interact.cpp [hunk 4]
```diff
@@ -847,15 +847,29 @@ bool veh_interact::update_part_requirements()
...
+                                    lift_strength - player_character.get_str() );
+    } else if( player_character.has_trait( trait_id( "BADBACK" ) ) && total_base_strength >= str &&
+               total_lift_strength < str ) {
+        str_suffix = string_format( _( "(Bad Back hindered by %d strength)\n" ),
```
Comment (actual-nh) [inline src/veh_interact.cpp:861]:
```suggestion
        str_suffix = string_format( _( "(Bad Back reduced usable strength by %d)\n" ),
```

File: src/veh_interact.cpp [hunk 5]
```diff
@@ -1806,6 +1820,22 @@ bool veh_interact::can_remove_part( int idx, const player &p )
...
     }
+
+    std::string str_suffix;
+    int lift_strength = player_character.get_lift_str();
+    int total_lift_strength = lift_strength + player_character.get_lift_assist();
+    int total_base_strength = player_character.get_str() + player_character.get_lift_assist();
+
+    if( player_character.has_trait( trait_id( "STRONGBACK" ) ) && total_lift_strength >= str &&
+        total_base_strength < str ) {
+        str_suffix = string_format( _( "(Strong Back helped with +%d strength)\n" ),
```
Comment (actual-nh) [inline src/veh_interact.cpp:1831]:
```suggestion
        str_suffix = string_format( _( "(Strong Back helped, giving +%d strength)\n" ),
```

File: src/veh_interact.cpp [hunk 6]
```diff
@@ -1806,6 +1820,22 @@ bool veh_interact::can_remove_part( int idx, const player &p )
...
+                                    lift_strength - player_character.get_str() );
+    } else if( player_character.has_trait( trait_id( "BADBACK" ) ) && total_base_strength >= str &&
+               total_lift_strength < str ) {
+        str_suffix = string_format( _( "(Bad Back hindered by %d strength)\n" ),
```
Comment (actual-nh) [inline src/veh_interact.cpp:1835]:
```suggestion
        str_suffix = string_format( _( "(Bad Back reduced usable strength by %d)\n" ),
```

**Review by actual-nh — COMMENTED** [review]
Looks good!

File: src/player.cpp [hunk 7]
```diff
@@ -1898,21 +1898,27 @@ player::wear( item_location item_wear, bool interactive )
...
+
 template <typename T>
 bool player::can_lift( const T &obj ) const
 {
     // avoid comparing by weight as different objects use differing scales (grams vs kilograms etc)
-    int str = get_str();
+    int str = get_lift_str();
     if( mounted_creature ) {
         const auto mons = mounted_creature.get();
         str = mons->mech_str_addition() == 0 ? str : mons->mech_str_addition();
```
Comment (BrettDong) [inline src/player.cpp:1919]: `str` could be altered here in the original code. So lift adjusted strength is not necessarily always based on `get_str()`. `player::get_lift_str()` should be changed to something like `player::adjust_life_str(int str)` that takes an integer `str` and calculates adjusted strength based on the integer parameter given.

Comment (Jamuro-g) [reply src/player.cpp:1918]: uploaded

File: src/player.cpp [hunk 8]
```diff
@@ -1898,21 +1898,26 @@ player::wear( item_location item_wear, bool interactive )
...
 {
     // avoid comparing by weight as different objects use differing scales (grams vs kilograms etc)
-    int str = get_str();
+    int str = adjust_lift_str( get_str() );
     if( mounted_creature ) {
         const auto mons = mounted_creature.get();
         str = mons->mech_str_addition() == 0 ? str : mons->mech_str_addition();
     }
     const int npc_str = get_lift_assist();
-    if( has_trait( trait_id( "STRONGBACK" ) ) ) {
```
Comment (BrettDong) [inline src/player.cpp:1911]: Thank you for updating. I'm sorry if I didn't explain clearly last time. `str` can be altered in `if(mounted_creature)` block before adjusting for `STRONGBACK`, so you need to follow the way the original code adjusted strength, that is, adjust strength after `if( mounted_creature )`.

File: src/veh_interact.cpp [hunk 9]
```diff
@@ -847,15 +847,29 @@ bool veh_interact::update_part_requirements()
...
     }
 
+    std::string str_suffix;
+    int lift_strength = player_character.get_lift_str();
+    int total_lift_strength = lift_strength + player_character.get_lift_assist();
+    int total_base_strength = player_character.get_str() + player_character.get_lift_assist();
+
+    if( player_character.has_trait( trait_id( "STRONGBACK" ) ) && total_lift_strength >= str &&
+        total_base_strength < str ) {
+        str_suffix = string_format( _( "(Strong Back helped, giving +%d strength)\n" ),
```
Comment (BrettDong) [inline src/veh_interact.cpp:857]: I think `\n` is not necessary here.

**Review by BrettDong — APPROVED** [review]

File: src/veh_interact.cpp (hunk 9 above)
Comment (Jamuro-g) [reply src/veh_interact.cpp:857]: i ll take a look, but i think without it any automatic linebreak ends up at the end of the description header.

Comment (BrettDong) [reply src/veh_interact.cpp]: I noticed there was a display bug so I made #47121

Comment (Jamuro-g) [reply src/veh_interact.cpp]: ah thank you for the fix :)

**Pull request merged**